lib/SILOptimizer/SILCombiner/SILCombinerCastVisitors.cpp

```diff
@@ -95,16 +96,120 @@ SILCombiner::visitRefToRawPointerInst(RefToRawPointerInst *rrpi) {
   return nullptr;
 }
 
-SILInstruction *SILCombiner::visitUpcastInst(UpcastInst *UCI) {
-  if (UCI->getFunction()->hasOwnership())
+namespace {
+
+/// A folder object for sequences of forwarding instructions that forward owned
+/// ownership. Is used to detect if we can delete the intermediate forwarding
+/// instructions without ownership issues and then allows the user to either
+/// delete all of the rest of the forwarding instructions and then replace front
+/// with a new value or set front's operand to a new value.
+class SingleBlockOwnedForwardingInstFolder {
+  SmallVector<SingleValueInstruction *, 4> rest;
+  SILCombiner &SC;
+  SingleValueInstruction *front;
+
+public:
+  SingleBlockOwnedForwardingInstFolder(
+      SILCombiner &SC, SingleValueInstruction *instructionToFold)
+      : SC(SC), front(instructionToFold) {
+    // If our initial instruction to fold isn't owned, set it to nullptr to
+    // indicate invalid.
+    if (SILValue(instructionToFold).getOwnershipKind() != OwnershipKind::Owned)
+      front = nullptr;
+  }
+
+  bool isValid() const { return bool(front); }
+
+  bool add(SingleValueInstruction *next) {
+    assert(isValid());
+    if (SILValue(next).getOwnershipKind() != OwnershipKind::Owned)
+      return false;
+
+    if (next->getSingleUse()) {
+      rest.push_back(next);
+      return true;
+    }
+
+    if (front->getParent() != next->getParent()) {
+      return false;
+    }
+
+    // Otherwise, since the two values are in the same block and we want to
+    // optimize only if our original value doesn't have any non-debug uses, we
+    // know that our value can only have a single non-debug use, the consuming
+    // user. So if we are not in that situation, bail.
+    if (!hasOneNonDebugUse(next))
+      return false;
+
+    rest.push_back(next);
+    return true;
+  }
+
+  /// Delete all forwarding uses and then RAUW front with newValue.
+  SingleValueInstruction *optimizeWithReplacement(SILValue newValue) && {
+    // NOTE: Even though after running cleanup rest, front now has its
+    // forwarding operand set to Undef, we haven't touched its result. So it is
+    // safe to RAUW.
+    cleanupRest();
+    SC.replaceValueUsesWith(front, newValue);
+    return nullptr;
+  }
+
+  /// Delete all forwarding uses and then set front's first operand to be \p
+  /// newValue.
+  SingleValueInstruction *optimizeWithSetValue(SILValue newValue) && {
+    cleanupRest();
+    assert(isa<SILUndef>(front->getOperand(0)));
+    front->setOperand(0, newValue);
+    SC.setUseValue(&front->getOperandRef(0), newValue);
     return nullptr;
+  }
+
+private:
+  /// From backwards -> forwards, for each instruction in rest, delete all of
+  /// its debug uses and then set its single remaining use to be SILUndef.
+  ///
+  /// This means that after this runs front's forwarding operand is now
+  /// SILUndef.
+  void cleanupRest() & {
+    // We process backwards -> forwards. This cleans up everything but the front
```

Comment:
> We process backwards -> forwards.

I can't parse this sentence. I think you mean we process top-down in the def-use chain

Reply:
yes

Reply:
I'll fix the comment.